File: TPs/TP02/Main.java
//Algoritmos e Estruturas de Dados 2 ----- Trabalho Prático 02
//Questão 01 ----- Modelagem em Java

import java.io.*;
import java.util.Scanner;

public class Main {

    public static void main(String[] args) throws Exception {

        Scanner sc = new Scanner(System.in);
        ColecaoRestaurantes colecao = ColecaoRestaurantes.lerCsv();

        int id;

        while ((id = sc.nextInt()) != -1) {

            for (int i = 0; i < colecao.getTamanho(); i++) {
                if (colecao.getRestaurantes()[i].getId() == id) {
                    System.out.println(
                        colecao.getRestaurantes()[i].formatar()
                    );
                    break;
                }
            }
        }

        sc.close();
    }
}

class Data {

    private int dia;
    private int mes;
    private int ano;

    public Data(int dia, int mes, int ano) {
        this.dia = dia;
        this.mes = mes;
        this.ano = ano;
    }

    public static Data parseData(String s) {
        int ano = Integer.parseInt(s.substring(0, 4));
        int mes = Integer.parseInt(s.substring(5, 7));
        int dia = Integer.parseInt(s.substring(8, 10));
        return new Data(dia, mes, ano);
    }

    public String formatar() {
        return String.format("%02d/%02d/%04d", dia, mes, ano);
    }

    public int getAno() {
        return ano;
    }
}

class Hora {

    private int hora;
    private int minuto;

    public Hora(int hora, int minuto) {
        this.hora = hora;
        this.minuto = minuto;
    }

    public static Hora parseHora(String s) {
        int hora = Integer.parseInt(s.substring(0, 2));
        int minuto = Integer.parseInt(s.substring(3, 5));
        return new Hora(hora, minuto);
    }

    public String formatar() {
        return String.format("%02d:%02d", hora, minuto);
    }
}

class Restaurante {

    private int id;
    private String nome;
    private String cidade;
    private int capacidade;
    private double avaliacao;
    private String[] tiposCozinha;
    private int faixaPreco;
    private Hora abertura;
    private Hora fechamento;
    private Data dataAbertura;
    private boolean aberto;

    public int getId() { return id; }
    public String getNome() { return nome; }

    public static Restaurante parseRestaurante(String linha) {

        String[] campos = linha.split(",");

        Restaurante r = new Restaurante();

        r.id = Integer.parseInt(campos[0]);
        r.nome = campos[1];
        r.cidade = campos[2];
        r.capacidade = Integer.parseInt(campos[3]);
        r.avaliacao = Double.parseDouble(campos[4]);

        r.tiposCozinha = campos[5].split(";");

        // faixa preço ($$$ → 3)
        r.faixaPreco = campos[6].length();

        // horário
        String[] partesHora = campos[7].split("-");
        r.abertura = Hora.parseHora(partesHora[0]);
        r.fechamento = Hora.parseHora(partesHora[1]);

        // data
        r.dataAbertura = Data.parseData(campos[8]);
        r.aberto = campos[9].equals("true");

        return r;
    }

    public String formatar() {

        String tipos = "[";
        for (int i = 0; i < tiposCozinha.length; i++) {
            tipos += tiposCozinha[i];
            if (i < tiposCozinha.length - 1) tipos += ",";
        }
        tipos += "]";

        String preco = "";
        for (int i = 0; i < faixaPreco; i++) preco += "$";

        return "["
                + id + " ## "
                + nome + " ## "
                + cidade + " ## "
                + capacidade + " ## "
                + avaliacao + " ## "
                + tipos + " ## "
                + preco + " ## "
                + abertura.formatar() + "-" + fechamento.formatar() + " ## "
                + dataAbertura.formatar() + " ## "
                + aberto
                + "]";
    }
}

class ColecaoRestaurantes {

    private int tamanho;
    private Restaurante[] restaurantes;

    public int getTamanho() {
        return tamanho;
    }

    public Restaurante[] getRestaurantes() {
        return restaurantes;
    }

    public void lerCsv(String path) throws Exception {

        BufferedReader br = new BufferedReader(new FileReader(path));

        restaurantes = new Restaurante[10000];
        tamanho = 0;

        br.readLine(); 

        String linha;
        while ((linha = br.readLine()) != null) {
            restaurantes[tamanho++] = Restaurante.parseRestaurante(linha);
        }

        br.close();
    }

    public static ColecaoRestaurantes lerCsv() throws Exception {
        ColecaoRestaurantes c = new ColecaoRestaurantes();
        c.lerCsv("/tmp/restaurantes.csv");
        return c;
    }
}

File: TPs/TP02/Main.c
// Algoritmos e Estruturas de Dados 2 ----- Trabalho Prático 02
// Questão 02 ----- Modelagem em C

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    int dia;
    int mes;
    int ano;
} Data;

Data parse_data(char* s) {
    Data d;
    sscanf(s, "%d-%d-%d", &d.ano, &d.mes, &d.dia);
    return d;
}

void formatar_data(Data* d, char* buffer) {
    sprintf(buffer, "%02d/%02d/%04d", d->dia, d->mes, d->ano);
}

typedef struct {
    int hora;
    int minuto;
} Hora;

Hora parse_hora(char* s) {
    Hora h;
    sscanf(s, "%d:%d", &h.hora, &h.minuto);
    return h;
}

void formatar_hora(Hora* h, char* buffer) {
    sprintf(buffer, "%02d:%02d", h->hora, h->minuto);
}

typedef struct {
    int id;
    char nome[200];
    char cidade[200];
    int capacidade;
    double avaliacao;

    char tiposCozinha[10][100];
    int numTipos;

    int faixaPreco;

    Hora abertura;
    Hora fechamento;

    Data dataAbertura;

    int aberto;
} Restaurante;

// converter $$$ → 3
int converter_preco(char* s) {
    return strlen(s);
}

Restaurante* parse_restaurante(char* linha) {

    Restaurante* r = (Restaurante*) malloc(sizeof(Restaurante));

    char tipos[200];
    char horario[50];
    char data[50];
    char abertoStr[10];

    sscanf(linha,
        "%d,%[^,],%[^,],%d,%lf,%[^,],%[^,],%[^,],%[^,],%s",
        &r->id,
        r->nome,
        r->cidade,
        &r->capacidade,
        &r->avaliacao,
        tipos,
        abertoStr,  
        horario,
        data,
        abertoStr
    );

    // tipos cozinha
    r->numTipos = 0;
    char* token = strtok(tipos, ";");
    while (token != NULL) {
        strcpy(r->tiposCozinha[r->numTipos++], token);
        token = strtok(NULL, ";");
    }

    // faixa preco
    r->faixaPreco = converter_preco(abertoStr);

    // horario
    char h1[10], h2[10];
    sscanf(horario, "%[^-]-%s", h1, h2);
    r->abertura = parse_hora(h1);
    r->fechamento = parse_hora(h2);

    // data
    r->dataAbertura = parse_data(data);

    r->aberto = (strcmp(abertoStr, "true") == 0);

    return r;
}

void formatar_restaurante(Restaurante* r, char* buffer) {

    char tipos[300] = "[";
    for (int i = 0; i < r->numTipos; i++) {
        strcat(tipos, r->tiposCozinha[i]);
        if (i < r->numTipos - 1) strcat(tipos, ",");
    }
    strcat(tipos, "]");

    char preco[10] = "";
    for (int i = 0; i < r->faixaPreco; i++) strcat(preco, "$");

    char abertura[10], fechamento[10], data[20];
    formatar_hora(&r->abertura, abertura);
    formatar_hora(&r->fechamento, fechamento);
    formatar_data(&r->dataAbertura, data);

    sprintf(buffer,
        "[%d ## %s ## %s ## %d ## %.1lf ## %s ## %s ## %s-%s ## %s ## %s]",
        r->id,
        r->nome,
        r->cidade,
        r->capacidade,
        r->avaliacao,
        tipos,
        preco,
        abertura,
        fechamento,
        data,
        r->aberto ? "true" : "false"
    );
}

typedef struct {
    int tamanho;
    Restaurante* restaurantes[10000];
} ColecaoRestaurantes;

void ler_csv_colecao(ColecaoRestaurantes* c, char* path) {

    FILE* f = fopen(path, "r");

    char linha[1000];

    fgets(linha, sizeof(linha), f); 

    c->tamanho = 0;

    while (fgets(linha, sizeof(linha), f)) {
        linha[strcspn(linha, "\n")] = 0;
        c->restaurantes[c->tamanho++] = parse_restaurante(linha);
    }

    fclose(f);
}

ColecaoRestaurantes* ler_csv() {
    ColecaoRestaurantes* c = (ColecaoRestaurantes*) malloc(sizeof(ColecaoRestaurantes));
    ler_csv_colecao(c, "/tmp/restaurantes.csv");
    return c;
}

int main() {

    ColecaoRestaurantes* c = ler_csv();

    int id;

    while (scanf("%d", &id) && id != -1) {

        for (int i = 0; i < c->tamanho; i++) {
            if (c->restaurantes[i]->id == id) {

                char buffer[1000];
                formatar_restaurante(c->restaurantes[i], buffer);
                printf("%s\n", buffer);

                break;
            }
        }
    }

    return 0;
}
